### ISSUE TITLE
Code review before publishing

### ISSUE THREAD

**Issue by Planeshifter** [validate-io/probability-array#1]

**Comment by kgryte** [comment]
Should we just check for `isArrayLike`, rather than `isArray` *or* `isTypedArrayLike`? Or are you wanting both strict: `!isArray && !isTypedArray`? 

Also would be good to include a precision test case; e.g., `.1 + .2 + .1 + .1 + .2 + .2 + .1 != 1.0`.

**Comment by Planeshifter** [comment]
Re: test case. Makes sense, will add one.
My rationale for the used check was that I wanted it to succeed for strings, if I recall correctly. Not sensible?

**Comment by kgryte** [comment]
Succeed for strings?

**Comment by Planeshifter** [comment]
Sorry. Should have been more precise. This if statement
```
if ( !isArray( value ) &&  !isTypedArrayLike( value ) ) {
eturn false;
}

**Comment by Planeshifter** [comment]
Question: What kind of objects pass `isTypedArrayLike` but fail `isTypedArray`?

**Comment by kgryte** [comment]
One with the following properties (https://github.com/validate-io/typed-array-like/blob/master/lib/index.js), but whose constructor does not match one of the known constructor names for typed arrays.

**Comment by Planeshifter** [comment]
Right. But is there anything in the wild where one would expect this? One instance I can think of is when you sub-class a typed array. 
Re: array-like check. Which solution do you prefer?

**Comment by kgryte** [comment]
Should we account for something like `[0.9999999999999999]`? Or in other words, is there a mechanism by which we can distinguish floating-point errors from a sum probability which does not equal unity?

**Comment by kgryte** [comment]
Re: array-like. Ditto for sub-classing an `array`. May be safer to do an explicit `isString` check.

**Comment by Planeshifter** [comment]
We could add
```javascript
	if ( len === 1 ) {
		return value[ 0 ] === 1 ? true : false;
	}
```
to catch your example?

**Comment by kgryte** [comment]
We could, but this is also problematic if `[.1 + .2 + .1 + .1 + .2 + .2 + .1]`.

**Comment by Planeshifter** [comment]
So then this would be better?
```
 if ( len === 1 ) {
        if ( typeof value[ 0 ] !== 'number' ) {
           return false;
        }
        return abs( value[ 0 ] - 1 ) <= FLOAT64_EPSILON  ? true : false;
 }
```

**Comment by kgryte** [comment]
This is effectively the same thing as currently done in the loop. So basically, we are full circle.

The additional type check when provided a typed array is unnecessary, but this strays toward making the module a generic which delegates to a particular implementation.

Also, what do you think about checking at each iteration? Meaning, right now, the entire array is traversed, but we could just as easily bail as soon as the sum exceeds `1+eps`.

Would be nice if we could sniff an integer typed array, as that would probably affect the implementation. Namely, no need for an `eps` check.

**Comment by Planeshifter** [comment]
True. ;)

I can certainly adapt the module so that for typed arrays, the unnecessary checks are removed. For integer typed-arrays, it would have been great if we had implemented what we discussed in terms of an `integer` validation module which exposes fcns to check for arrays, typed-arrays, matrices... As of now, no `isIntegerTypedArray` module exists. :(

**Comment by kgryte** [comment]
Agreed. And currently, we have no way of determining whether a subclassed typed array descends from an integer typed array. 

Although, even if we could check for an integer array, we would take a perf hit by cycling through the array twice, rather than combining operations in a single `for` loop. Ah, and now I am recalling, we wanted to be able to compile validation functions. Yes, this would be good for the current use case.

For the case of a plain `array`, in the interests of perf, we could have just assumed a `Float64` array.
